### ISSUE TITLE
Resolve container hostname to Rancher network address

### ISSUE THREAD

**Issue by stephan2012** [rancher/rancher#5499]
Our containerized application (Oracle Weblogic, Managed Server) registers itself to a central hub server (Admin Server) with its IP address. It does this by taking the hostname(2) followed by gethostbyname(2) on it. Since the hostname is recorded in /etc/hosts with the 172.17.x.x address, the Managed Server registers with an IP address that the Admin Server cannot reach.

It'd be better if the IP address from the rancher network (10.42.x.x) is recorded in /etc/hosts, maybe optionally.

**Rancher Version:** v1.1.1

**Docker Version:** 1.11.2

**OS and where are the hosts located? (cloud, bare metal, etc):** RHEL 7.2 on XenCenter VM

**Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)** Multi-node, internal DB

**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)** Cattle

**Comment by stephan2012** [comment]
Setting the hostname to the container name does not help either because Docker then records the 172.17.x.x address in /etc/hosts while a DNS lookup returns the 10.42.x.x address.

**Comment by henrypfhu** [comment]
I have similar issue, in the Rancher managed docker instance, hostname is resolved to docker 172.17.x.x network ip address in /etc/hosts configuration file.

That will cause problem, when application use hostname to retrieve IP address and register this IP to some SOA registry center. That will cause bad issue, the client will get the wrong IP (docker 172.17.x.x) and use this wrong IP to access the remote service. If these two Rancher managed docker run in two different HOST, that will cause the problem. Client can not access Remote Services located in another HOST rancher docker instance.

It's too bad.

I know it's docker engine's problem to cause such bad thing.

I don't know whether Rancher can help resolve this bad issue?!

**Comment by henrypfhu** [comment]
If you use command 'ping `hostname`' in the Rancher Managed Docker instance, you will know what I'm saying..

**Comment by henrypfhu** [comment]
+1

**Comment by leodotcloud** [comment]
You can use rancher-metadata service to figure out the primary IP address.
`curl 'http://rancher-metadata/2015-12-19/self/container'`

**Comment by stephan2012** [comment]
Thanks for the pointer to the Rancher metadata service.

Rancher IP address is configured after the container was started or is there a possibility to pass metadata as environment variables to a container somehow? I mean to not pull metadata from inside the container. Otherwise special logical would be required just in case a container images is deployed on Rancher…

**Comment by stefli** [comment]
Any update?

**Comment by stefli** [comment]
@stephan2012 I found a solution to obtain the rancher managed ip by using entrypoint.sh. In the entrypoint.sh,we can the container's primary ip(10.42.x.x) and hostname, then rewrite them to /etc/hosts. All works fine, code likes:
`#!/bin/bash

CONTAINER_IP=$(curl http://rancher-metadata/latest/self/container/primary_ip)
CONTAINER_NAME=$(hostname)

echo "127.0.0.1 localhost" > /etc/hosts
echo "$CONTAINER_IP $CONTAINER_NAME" >> /etc/hosts

java -jar xx.jar 

exec "$@"`

**Comment by stephan2012** [comment]
@stefli Thnaks for your hint. I worked around it by using the Rancher Metadata service. However, I prefer to keep container images agnostic to the orchestration and network solution.

**Comment by stefli** [comment]
@stephan2012 Could you please tell me your solution, thanks.
Change your application source code to obtain primary ip or any other configuration in Rancher UI?

**Comment by stephan2012** [comment]
Well… In the Docker templates for Weblogic (https://github.com/oracle/docker-images/) are some scripts. I've adjusted docker-images/OracleWebLogic/samples/1221-domain/container-scripts/createServer.sh to include

```
# Rancher Managed Network special logic
RANCHERIP=$( curl -s --proxy '' 'http://rancher-metadata/2015-12-19/self/container/primary_ip' )
if [[ -n "$RANCHERIP" && "$RANCHERIP" != "Not Found" ]]
then
    export NM_HOST="$RANCHERIP"
    export MS_HOST="$RANCHERIP"
fi
```

However, for WLS this is not the full solution because after startup it struggles with #5761 and WLS does not like setting the hostname to the container name (maybe too long, maybe the underscores).

So, this is neither generic nor orchestration agnostic.

**Comment by leodotcloud** [comment]
@stefli @henrypfhu @stefli Can you please try the latest release of Rancher? The networking model has changed completely and you can read the 10.42.x.x IP address directly from eth0. There won't be two IP addresses for containers with 'managed' network option. And hostname of the container will resolve the Rancher IP address.

@stephan2012 If this resolves, can you close the issue?

**Comment by stephan2012** [comment]
Hi @leodotcloud, I'll check again but can do this only at the beginning of 2017.

**Comment by vigohe** [comment]
@leodotcloud I'm doing the same hack as @stephan2012 but for spring-cloud for this version did work without issues :

> Rancher	v1.1.0-dev5
Cattle	v0.164.1
User Interface	v1.1.10
Rancher Compose	v0.8.4
Docker  1.10.3

but now I'm playing with and the hack did not work.

> Rancher	v1.2.1
Cattle	v0.174.8
User Interface	v1.2.39
Rancher Compose	v0.12.1
Docker  1.10.3


I notice that Pinging any 10.42.x.x IP from within container in Rancher v1.1.0-dev5 does work, but with Cattle v1.2.1 is not working for me. Just works the ones I have link or are in the same host or are in the stack.

Why I can ping a 10.42.x.x IP of another host? It's a bug of rancher v1.2.1? Does v1.2.2 solve it?

**Comment by vigohe** [comment]
I can confirm that with rancher v1.2.2 pinging 10.42.x.x IP works.

**Comment by leodotcloud** [comment]
@stephan2012 From within the container (say 3ec29fb833b6), `ping 3ec29fb833b6` resolves to the `10.42.x.y` IP address. But from a different container (fc60f3450563) you try `ping 3ec29fb833b6` it doesn't resolve. Rancher doesn't populate it's DNS server based on the container IDs.

Also please see my previous comment and see if that resolves the issue.

**Comment by superseb** [comment]
As this originates from old network implementation to new one, please reopen if you are still having issues with resolving addresses. Thanks
